radix-engine/src/track/db_key_mapper.rs

```diff
@@ -1,71 +1,147 @@
 use radix_engine_common::data::scrypto::{scrypto_decode, ScryptoDecode};
-use radix_engine_interface::crypto::{hash, Hash};
+use radix_engine_interface::crypto::hash;
 use radix_engine_interface::types::{ModuleId, NodeId, SubstateKey};
 use radix_engine_store_interface::interface::{
-    DbPartitionKey, DbSortKey, PartitionEntry, SubstateDatabase,
+    DbPartitionKey, DbSortKey, DbSubstateValue, SubstateDatabase,
 };
 use sbor::rust::prelude::*;
+use utils::{combine, copy_u8_array};
 
 /// A mapper between the business RE Node / Module / Substate IDs and database keys.
 pub trait DatabaseKeyMapper {
     /// Converts the given RE Node and Module ID to the database partition's key.
+    /// Note: contrary to the sort key, we do not provide the inverse mapping here (i.e. if you
+    /// find yourself needing to map the database partition key back to RE Node and Module ID, then
+    /// you are most likely using the "partition vs sort key" construct in a wrong way).
     fn to_db_partition_key(node_id: &NodeId, module_id: ModuleId) -> DbPartitionKey;
 
     /// Converts the given Substate key to the database's sort key.
     fn to_db_sort_key(key: &SubstateKey) -> DbSortKey;
+
+    /// Converts the given database's sort key to a Substate key.
+    fn from_db_sort_key(db_sort_key: &DbSortKey) -> SubstateKey;
 }
 
 /// A [`DatabaseKeyMapper`] tailored for databases which cannot tolerate long common prefixes
 /// among keys (for performance reasons). In other words, it spreads the keys "evenly" (i.e.
-/// pseudo-randomly) across the key space.
-/// For context: our actual use-case for it is the Jellyfish Merkle Tree.
+/// pseudo-randomly) across the key space. For context: our use-case for this is the Jellyfish
+/// Merkle Tree.
+///
+/// This implementation is the actual, protocol-enforced one, to be used in public Radix networks.
+///
+/// This implementation achieves the prefix-spreading by:
+/// - using a (long, ~100% unique) hash instead of plain RE Node and Module ID (please note that it
+///   makes this mapping effectively irreversible);
+/// - using a (shorter, but hard to crack) hash prefix for Substate key.
 pub struct SpreadPrefixKeyMapper;
 
 impl DatabaseKeyMapper for SpreadPrefixKeyMapper {
     fn to_db_partition_key(node_id: &NodeId, module_id: ModuleId) -> DbPartitionKey {
```

Comment:
I know we've left the call now - but I think I do want there to be some way to invert this...

Notably if we need to migrate to a new key-ing paradigm (say, a move to verkle trees), being able to migrate all the data is super useful. Having _some way_ of inverting these would prove useful no doubt.

Ditto - it would be _super useful_ for trustless clients (eg wallets) to know the actual address of something -- or indeed at Xi'an when we're using merkle trees a lot - having this data would be really useful.

Maybe we add the "own address" to TypeInfo @talekhinezh ?

Reply:
Remember a separate index could always be used for both trustless clients and large tree migrations.

I'd prefer optimizing and keeping this execution store for it's main purpose, ie represents proveable state on top of which transactions can be deterministically run on top off.

Other things like querying/traversing can be much better served by separate indices/data structures.